proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -207,64 +373,261 @@ lemma obj_at_setObject2:
+    the new value; we get to "reuse" the existing PSpace properties. \<close>
+lemma same_size_obj_at'_set_obj'_iff:
+  fixes obj :: "'a :: pspace_storable"
+  assumes "obj_at' (\<lambda>old_obj :: 'a. objBits old_obj = objBits obj) ptr s"
+  shows "obj_at' P ptr (set_obj' ptr obj s) = P obj"
+  apply (rule iffI)
+   apply (prop_tac "ko_at' obj ptr (set_obj' ptr obj s)")
+    apply (clarsimp simp: obj_at'_def projectKO_eq project_inject)
+   apply (clarsimp simp: obj_at'_def)
+  using assms
+  apply (fastforce simp: obj_at'_def inj_def projectKO_eq project_inject objBits_def)
+  done
+
+lemma tcb_at'_obj_at'_set_obj'[unfolded injectKO_tcb]:
+  assumes "P (tcb :: tcb)"
+      and "tcb_at' ptr s"
+  shows "obj_at' P ptr (set_obj' ptr tcb s)"
+  using assms
+  apply (clarsimp simp: objBits_def objBitsKO_def inj_def
+                        same_size_obj_at'_set_obj'_iff[where 'a=tcb, simplified])
+  done
+
+\<comment>\<open> Keeps a generic @{term obj_at'} (rather than a specific @{term "obj_at' (\<lambda>_. True)"}) to match
+    in more simp contexts. \<close>
+lemma tcb_obj_at'_set_obj'_iff:
+  fixes tcb :: tcb
+    and P Q :: "tcb \<Rightarrow> bool"
+  shows "obj_at' P p s \<Longrightarrow> obj_at' Q p (set_obj' p tcb s) = Q tcb"
+  apply (rule same_size_obj_at'_set_obj'_iff)
+  apply (clarsimp simp: objBits_simps obj_at'_def)
+  done
+
+lemmas tcb_obj_at'_pred_tcb'_set_obj'_iff =
+  tcb_obj_at'_set_obj'_iff[where Q="test o proj o tcb_to_itcb'" for test proj,
+                                 simplified objBits_simps o_def, simplified,
+                                 folded pred_tcb_at'_def]
+
+lemma same_size_ko_wp_at'_set_ko'_iff:
+  assumes "ko_wp_at' (\<lambda>old_ko. objBitsKO old_ko = objBitsKO ko) ptr s"
+  shows "ko_wp_at' P ptr (set_ko' ptr ko s) = P ko"
+  apply (rule iffI)
+   apply (clarsimp simp: ko_wp_at'_def)
+  using assms
+  apply (clarsimp simp: ko_wp_at'_def)
+  apply (erule ps_clear_domE)
+  apply clarsimp
+  apply blast
+  done
+
+\<comment>\<open> Moves the @{term ksPSpace_update} to the top. \<close>
+lemma unfold_set_ko':
+  "set_ko' ptr ko s = ksPSpace_update (\<lambda>ps. ps(ptr := Some ko)) s"
+  by clarsimp
+
+lemma ko_wp_at'_set_ko'_distinct:
+  assumes "ptr \<noteq> ptr'"
+          "ko_wp_at' \<top> ptr' s"
+  shows "ko_wp_at' P ptr (set_ko' ptr' ko s) = ko_wp_at' P ptr s"
+  using assms
+  apply (clarsimp simp: ko_wp_at'_def)
+  apply (rule iffI; clarsimp)
+   apply (erule ps_clear_domE)
+   apply clarsimp
+   apply blast
+  apply (erule ps_clear_domE)
+  apply clarsimp
+  apply blast
+  done
 
-lemma updateObject_tcb_eta:
-  "updateObject (v :: tcb) = updateObject_default v"
-  by ((rule ext)+, simp)
+lemma obj_at'_set_obj'_distinct:
+  assumes "p \<noteq> p'"
+          "obj_at' Q p' s"
+  shows "obj_at' P p (set_ko' p' ko s) = obj_at' P p s"
+  using assms apply -
+  apply (rule iffI; fastforce simp: obj_at'_def projectKO_eq project_inject ps_clear_upd)
```

Comment:
`apply` should be on its own line

Reply:
To avoid the whole `using assms apply -` thing, we could rewrite the lemma statement into non-Isar from (no assumes/shows). The proof doesn't make use of any of the Isar features and the statement might actually be more readable.